### PR TITLE
Change Dive default and PD plot error handling in WIT 

### PR DESCRIPTION
**Pull request by jameswex**
* Motivation for features / changes

Dive positioning default is wrong for binary classification models
PD plot error handling stops future PD plots from being rendered.

* Technical description of changes

Change which axis does the scattering in this case, for better visuals
When PD plot generation fails, use a new error callback to clear out the progress spinner and remove the feature name from the pending charts list.
Do better parsing of feature value indices on python side
Various visual fixes with confusion matrix total row/column text size/color, PD plot tooltip precision, tab text spacing.

* Screenshots of UI changes

initialview

* Detailed steps to verify changes work correctly (as executed by you)

Ran demos
Ran python unit tests
Forced errors in PD plot generation and saw correct handling of the errors

* Alternate designs / implementations considered

### CODE REVIEW

**Comment by jameswex** [comment]
@tolga-b please review

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 1]
```diff
@@ -4387,6 +4390,9 @@ <h2>Show similarity to selected datapoint</h2>
             // show toast with the error
             this.showToast_(result.value.error);
             this.$.spinner.hidden = true;
+            if (errorFn != null) {
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: may want to be careful with this -- if you change this to `!==`, it will start to fail.
```
> null == undefined
< true
> null === undefined
< false
```

How about default arg to errorFn in L4384? (`errorFn= () => {}`)

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 2]
```diff
@@ -5065,11 +5079,14 @@ <h2>Show similarity to selected datapoint</h2>
         if (!this.local) {
           const url = tf_backend.addParams(
             '/data/plugin/whatif/infer_mutants', urlParams);
-          const chartMakerCallback = function(result){
+          const chartMakerCallback = result =>
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: nit: for legibility, if an arrow function is more than one line long, please use braces.
i.e., 
```js
// A bit hard to read
foo = result => 
    iAmGoingToReturnSomething(
       result);
// To below
foo = result => {
    return iAmGoingToReturnSomething(
       result);
};
```

Here and L5086

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py [hunk 3]
```diff
@@ -79,14 +79,11 @@ def convert_pattern_to_indices(pattern):
...
-        try:
-          if '-' in piece:
-            lower, upper = [int(x.strip()) for x in piece.split('-')]
-            indices.extend(range(lower, upper + 1))
-          else:
-            indices.append(int(piece.strip()))
-        except ValueError as e:
-          raise common_utils.InvalidUserInputError(e)
+        if '-' in piece:
+          lower, upper = [int(x.strip()) for x in piece.split('-')]
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/utils/inference_utils.py]: Please add second argument to `split`. 
```py
# Does this blow up if piece contains more than one '-'?
lower, upper = [int(x.strip()) for x in piece.split('-',)]

# Kinda better
lower, upper = [int(x.strip()) for x in piece.split('-', 1)]
```

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py [hunk 4]
```diff
@@ -97,7 +94,12 @@ def convert_pattern_to_indices(pattern):
     # By default, there are no specific user-requested feature indices.
     self.feature_indices = []
     if feature_index_pattern:
-      self.feature_indices = convert_pattern_to_indices(feature_index_pattern)
+      try:
+        self.feature_indices = convert_pattern_to_indices(
+          feature_index_pattern)
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/utils/inference_utils.py]: super nit: +2 whitespace as per Google's Python style guide.

File: tensorboard/plugins/interactive_inference/utils/inference_utils_test.py [hunk 5]
```diff
@@ -456,6 +456,36 @@ def test_convert_predict_response_classification(self):
...
+        examples=[],
+        num_mutants=0,
+        feature_index_pattern='3, 1')
+    self.assertEqual([1, 3], viz_params.feature_indices)
+    viz_params = inference_utils.VizParams(
+        x_min=1,
+        x_max=10,
+        examples=[],
+        num_mutants=0,
+        feature_index_pattern='0-')
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/utils/inference_utils_test.py]: can we add a bit more test around corner case? 
i.e.,
- `feature_index_pattern='0-1-2'`
- `feature_index_pattern='3-1'`
- `feature_index_pattern='1-Foo'`

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 2 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py (hunk 4 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/utils/inference_utils.py]: done

File: tensorboard/plugins/interactive_inference/utils/inference_utils_test.py (hunk 5 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/utils/inference_utils_test.py]: added more tests

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py (hunk 3 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/utils/inference_utils.py]: added argument and made sure that with more than one - it will treat as a parse error. see tests

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 1 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: added default arg. kept this null check  in place. we use the == and != syntax around WIT

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-confusion-matrix.html [hunk 6]
```diff
@@ -38,6 +39,14 @@
         text-align: left;
         font-size: 10px;
       }
+      .total-label {
+        text-align: left;
```
[the remark this answers is not in the document]
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-confusion-matrix.html]: we haven't sorted in WIT at all. should we sort all these in a separate PR that just handles this large change? or only sort what we encounter when we do PRs like this?

**Comment by jameswex** [comment]
@stephanwlee thanks, responded

**Comment by wchargin** [comment]
@jameswex: When you merge PRs, could you please make sure that the
commit message matches the PR description? For instance, this PR was
merged as 0c518f3b00bb6edaab0e9abdbef084751aa82bcf with this message:

> ```
> * Add flags to TBContext
> 
> * Add flags to TBContext in application
> 
> * Add flags to tbcontext
> 
> * gitignore changes
> 
> * fix merge issue
> 
> * new default
> 
> * fix pd plot error handling
> 
> * fix issues
> 
> * review comments
> ```

which is not really correct (this didn’t add any flags to `TBContext`)
and is much less helpful than your nicely formatted PR description. :-)

It’s annoying that GitHub’s UI makes it so easy to do the wrong thing
here, but given that it does we’d appreciate it if you could copy the
Markdown source from the PR description and paste it into the “commit
description” text box prior to merging. Thanks!
